src/actions/plugins/leap/LeapAction.ts

```diff
@@ -0,0 +1,139 @@
+import { Mode } from '../../../mode/mode';
+import { BaseCommand, RegisterAction } from '../../base';
+import { Position } from 'vscode';
+import { VimState } from '../../../state/vimState';
+import { configuration } from '../../../configuration/configuration';
+import { LeapSearchDirection, createLeap } from './leap';
+import { getMatches, generateMarkerRegex, generatePrepareRegex } from './match';
+import { StatusBar } from '../../../statusBar';
+import { Marker } from './Marker';
+import { VimError, ErrorCode } from '../../../error';
+
+@RegisterAction
+export class LeapPrepareAction extends BaseCommand {
+  modes = [Mode.Normal];
+  keys = [
+    ['s', '<character>'],
+    ['S', '<character>'],
+  ];
+
+  public override doesActionApply(vimState: VimState, keysPressed: string[]) {
+    return super.doesActionApply(vimState, keysPressed) && configuration.leap;
+  }
+
+  public override async exec(cursorPosition: Position, vimState: VimState): Promise<void> {
+    if (!configuration.leap) return;
+
+    if (this.keysPressed[1] === '\n') {
+      this.execRepeatLastSearch(vimState);
+    } else {
+      await this.execPrepare(cursorPosition, vimState);
+    }
+  }
+
+  private async execPrepare(cursorPosition: Position, vimState: VimState) {
+    const direction = this.getDirection();
+    const firstSearchString = this.keysPressed[1];
+
+    const leap = createLeap(vimState, direction, firstSearchString);
+    vimState.leap = leap;
+    vimState.leap.previousMode = vimState.currentMode;
+
+    const matches = getMatches(
+      generatePrepareRegex(firstSearchString),
+      direction,
+      cursorPosition,
+      vimState.document
+    );
+
+    vimState.leap.createMarkers(matches);
+    vimState.leap.showMarkers();
+    await vimState.setCurrentMode(Mode.LeapPrepareMode);
+  }
+
+  private execRepeatLastSearch(vimState: VimState) {
+    if (vimState.leap?.leapAction) {
+      vimState.leap.isRepeatLastSearch = true;
+      vimState.leap.direction = this.getDirection();
+      vimState.leap.leapAction.fire();
+    } else {
+      StatusBar.displayError(vimState, VimError.fromCode(ErrorCode.LeapNoPreviousSearch));
+    }
+  }
+
+  private getDirection() {
+    return this.keysPressed[0] === 's' ? LeapSearchDirection.Backward : LeapSearchDirection.Forward;
+  }
+}
+
+@RegisterAction
+export class LeapAction extends BaseCommand {
+  modes = [Mode.LeapPrepareMode];
+  keys = ['<character>'];
+  override isJump = true;
+  private vimState!: VimState;
```

Comment:
can you get rid of the `!`?

2 suggestions for how to do it:
1. Create a handler class with the vimState as constructor parameter, and use this class.
2. Pass the vimState from function to another as parameter.

Reply:
Haha thank you for the advice
As you said, this code-style problem

Reply:
Yeah I'd prefer `vimState` is passed explicitly to each method